### ISSUE TITLE
DESCRIBE SCHEMA for alternator table in restore context

### ISSUE THREAD

**Issue by Michal-Leszczynski** [scylladb/scylladb#19112]
Alternator keyspaces/tables can have invalid names from CQL POV, e.g. `alternator_Tab_le-With1.da_sh2-aNd.d33ot.-` is a valid name in alternator. AFAIK it's not possible to create a keyspace with such name via CQL - it is only possible with alternator API.

This wasn't a problem when schema was backed up/restored via sstables, as this path didn't go through CQL, but starting from Scylla 6.0 schema will be backed up/restored via the output of DESCRIBE SCHEMA WITH INTERNALS. The output of DESC SCHEMA contains correct alternator keyspace/table names, e.g. `CREATE KEYSPACE "alternator_Tab_le-With1.da_sh2-aNd.d33ot.-"`, but those CQL statements can't be used to recreate schema via CQL because they don't pass name validation.

In general, I'm not even sure that this name validation is the only problem, because creating table via cql and then trying to access it via alternator API might fail in different places. So what is the supported way of backing up/restoring schema of alternator table?

@nyh @Jadw1 can you advise on that? Is there some workaround that I'm missing?

@tzach @mykaul this looks like another regression for SM and Scylla 6.0 compatibility.

**Comment by nyh** [comment]
You are right, CQL and DynamoDB API do have different limitations on allowed table names. The differences are:
1. The dash (-) and dot (.) characters are allowed by DynamoDB but not by CQL
2. DynamoDB allows table names up to 255 characters long (Alternator actually limits this to 222).

When we started Alternator, and even as recently as a few months ago, the manager and product consensus was that users should not be allowed to mix CQL and DynamoDB API - not in the same cluster but most definitely not on the same table: You were not supposed to access your Alternator tables through CQL, or vice versa. This is why even if the Alternator table has a name which isn't legal for CQL, we used this name as-is as the table name; If we wanted to support CQL on Alternator tables, we should have "encoded" the illegal names in the legal characters (e.g., something analogous to https://en.wikipedia.org/wiki/Punycode). But we didn't.

But I would like to echo @Michal-Leszczynski's last point: It is not a good idea to create an Alternator table using CQL commands. **It's not just the name** - It is not guaranteed that an Alternator's `CreateTable` can be emulated by a CQL `CREATE TABLE`. In particular, besides the name, we have:

1. Alternator tables can also have materialized views. Those have even more bizarre names (including characters like "!") and I'm pretty sure a CREATE MATERIALIZED VIEW can't create them.
2. Alternator's materialized view allows two key columns, something which is not allowed by CREATE MATERIALIZED VIEW (https://github.com/scylladb/scylladb/issues/6714).
3. The DynamoDB API has the notion of "tags" on a table, and Alternator stores them as "extentions" on the schema. The codebase supports this feature well, but I don't think that CQL's DESCRIBE TABLE nor CREATE TABLE can read or write them.
4. In the future (IIRC, this doesn't happen today), Alternator tables might use new column types that aren't supported by CQL, or store things it needs to remember from the table creation command in system tables which CQL doesn't use.

It sounds to me that if we have Alternator tables, the safest approach is to list them (`ListTables`) and recreate them using Alternator, not CQL.

That being said, if we decide it's critical, there is something we can do if we really want to do this via CQL:
One thing we could do is to have `DESC TABLE` notice that this is an Alternator table, and add a flag like "WITH ALTERNATOR" or something, and the "CREATE TABLE" will, when given a "WITH ALTERNATOR", call the Alternator code (or some hacky compatibility code) to create the table, including using different rules for the name.

**Comment by kostja** [comment]
@nyh why it was decided to use dots in the first place? We could have escape-encoded the dots, a pretty arbitrary escape-encoding scheme would do. Perhaps we can still make the change and fix the problem for all new installs?

**Comment by kostja** [comment]
From grammar point of view we can work with dots and dashes just fine if we always quote our identifiers in DESCRIBE. The issue, however, is that dots and dashes are not allowed in identifiers in SQL, so external multi-vendor tools will hiccup on these names.

**Comment by kostja** [comment]
> You are right, CQL and DynamoDB API do have different limitations on allowed table names. The differences are:
> 
> 1. The dash (-) and dot (.) characters are allowed by DynamoDB but not by CQL
> 2. DynamoDB allows table names up to 255 characters long (Alternator actually limits this to 222), CQL allows only 48 (and wastes some of them on the prefix `alternator_`).
> 
> When we started Alternator, and even as recently as a few months ago, the manager and product consensus was that users should not be allowed to mix CQL and DynamoDB API - not in the same cluster but most definitely not on the same table: You were not supposed to access your Alternator tables through CQL, or vice versa. This is why even if the Alternator table has a name which isn't legal for CQL, we used this name as-is as the table name; If we wanted to support CQL on Alternator tables, we should have "encoded" the illegal names in the legal characters (e.g., something analogous to https://en.wikipedia.org/wiki/Punycode). But we didn't.
> 
> But I would like to echo @Michal-Leszczynski's last point: It is not a good idea to create an Alternator table using CQL commands. **It's not just the name** - It is not guaranteed that an Alternator's `CreateTable` can be emulated by a CQL `CREATE TABLE`. In particular, besides the name, we have:
> 
> 1. Alternator tables can also have materialized views. Those have even more bizarre names (including characters like "!") and I'm pretty sure a CREATE MATERIALIZED VIEW can't create them.
> 2. Alternator's materialized view allows two key columns, something which is not allowed by CREATE MATERIALIZED VIEW ([Document better the restriction of a view to have only one new key column #6714](https://github.com/scylladb/scylladb/issues/6714)).
> 3. The DynamoDB API has the notion of "tags" on a table, and Alternator stores them as "extentions" on the schema. The codebase supports this feature well, but I don't think that CQL's DESCRIBE TABLE nor CREATE TABLE can read or write them.
> 4. In the future (IIRC, this doesn't happen today), Alternator tables might use new column types that aren't supported by CQL, or store things it needs to remember from the table creation command in system tables which CQL doesn't use.
> 
> It sounds to me that if we have Alternator tables, the safest approach is to list them (`ListTables`) and recreate them using Alternator `CreateTable`, not CQL.
> 
> That being said, if we decide it's critical, there is something we can do if we really want to do this via CQL: One thing we could do is to have `DESC TABLE` notice that this is an Alternator table, and add a flag like "WITH ALTERNATOR" or something, and the "CREATE TABLE" will, when given a "WITH ALTERNATOR", call the Alternator code (or some hacky compatibility code) to create the table, including using different rules for the name.

Sorry I just noticed your long comment.
I disagree with your conclusions.

I think everything we do in Alternator should map to a CQL feature. Otherwise we get an explosion of features, since e.g. backup/restore or CDC or tablts have to be patched to work with the alternator. Any rule can have exceptions, but with names specifically, I'd say that it would be good to be able to create, drop, backup, restore alternator objects from CQL.

**Comment by nyh** [comment]
> Sorry I just noticed your long comment. I disagree with your conclusions.
> 
> I think everything we do in Alternator should map to a CQL feature. Otherwise we get an explosion of features, since e.g. backup/restore or CDC or tablts have to be patched to work with the alternator. Any rule can have exceptions, but with names specifically, I'd say that it would be good to be able to create, drop, backup, restore alternator objects from CQL.

One of the options I raised above was, indeed, to add a flag "WITH ALTERNATOR" (or similar) to the CQL "CREATE TABLE" and "CREATE MATERIALIZED VIEW".

But I think the blanket statement that "everything we do in Alternator should map to a CQL feature" is very problematic, and won't be easy to implement in a good way. If we want to remain 100% compatible with both CQL and the DynamoDB API (which I spent a good deal of time and effort to ensure), there are may features which are "similar" in spirit between CQL and DynamoDB, but differ in details. The data types are different, the filters are different, and much much more.

For this reaso, today Alternator requests are not translated into CQL, but into "internal" function calls to perform writes, queries, to create tables, and so on - plus Alternator-specific code (e.g., the filtering is done differently). This made the CQL code simpler (e.g., no need for CQL filtering to support **also** DynamoDB-style filtering) and allowed us to develop near-100% DynamoDB compatibility is a reasonable time without huge refactoring of the CQL code. I still think it was the correct decision at the time. If you want to reconsider this decision, we can do this. I could create a design document of what it would take to have full support for Alternator tables and DynamoDB-API requests from CQL. But I don't think it's relevant to **this** short-term issue.

**Comment by Michal-Leszczynski** [comment]
We should get this thing going somewhere.

AFAIU there are two directions:
- make it safe to use CQL for operating on alternator tables
- change SM backup/restore procedure, so that it uses alternator API instead of CQL where CQL is unsafe to use

Let's think about the second one.
I understand that using regular Scylla REST API is safe for alternator tables, so the problem is just about CQL (please correct me if I'm wrong).

For backup, CQL is used only to:
- check which tables are views (optional)
- backup `DESCRIBE SCHEMA WITH INTERNALS`

SM can check views with the `ListTable` and `DescribeTable` alternator API.

SM can also backup create table inputs from `DescribeTable` alternator API alongside `DESC SCHEMA WITH INTERNALS` as well. IIUC, SM would still need `DESC SCHEMA WITH INTERNALS` output for recreating roles, permissions, service levels, etc...

For restore, CQL is used for:
- applying described schema
- altering tables tombstone gc mode
- querying table version
- dropping/recreating views

As mentioned above, applying described schema would be a combination of backed up create table inputs from `DescribeTable` alternator API and all remaining entities from `DESC SCHEMA WITH INTERNALS`.
The question here would be: can SM restore tables just from the output of `DescribeTable` alternator API and expect that it is complete and safe for restore? What about things like dropped and re-added columns which are handled by `DROP ... USING TIMESTAMP` CQL statements? It doesn't seem like alternator `CreateTable` API can handle that. Would it be safe to create table with alternator API and the apply those `DROP ... USING TIMESTAMP` statements from `DESC SCHEMA WITH INTERNALS` output?

IIUC, alternator tables tombstone gc mode can be altered via CQL (please correct me if I'm wrong).
If not, SM can probably also use scylla REST API which allows to disable/enable table tombstone gc.

IIUC, querying table version from CQL should also be safe (SM needs it for creating a path to table upload dir).

Dropping and recreating views can be done with the `UpdateTable` alternator API.

Some additional consideration is that right now SM stores just CQL username and password needed for accessing cluster via CQL. Authenticating looks a little bit different in alternator API, as it requires salted hash instead of the actual password ([docs](https://opensource.docs.scylladb.com/stable/alternator/compatibility.html#authentication-and-authorization)). This shouldn't be a big deal, as we could add a flag like `--alternator-password` when adding cluster to SM (or add this flag per task definition in the spirit of https://github.com/scylladb/scylla-manager/issues/4409).

@nyh how does this look from your perspective? Does it seem like SM has all the tools to support alternator backup/restore or are we missing something here?

@tzach I can also prepare more detailed design doc for supporting alternator restore, but this would require some prioritization.

**Comment by mykaul** [comment]
@swasik - we need the team to help with this - we should improve the overall Alternator backup/restore experience.

**Comment by swasik** [comment]
@mykaul we discussed it shortly yesterday with @ScyllaPiotr. We believe that making sure that backup/restore works correctly and covering it by proper tests is of high priority from the perspective of reliability and data safety.

Now that we have the scope for Streams GA approved so we completed first stage of the project, our proposal is to put higher priority on it before starting to implement Streams gaps. What do you think?

**Comment by mykaul** [comment]
> [@mykaul](https://github.com/mykaul) we discussed it shortly yesterday with [@ScyllaPiotr](https://github.com/ScyllaPiotr). We believe that making sure that backup/restore works correctly and covering it by proper tests is of high priority from the perspective of reliability and data safety.
> 
> Now that we have the scope for Streams GA approved so we completed first stage of the project, our proposal is to put higher priority on it before starting to implement Streams gaps. What do you think?

Makes sense, thanks.

**Comment by tzach** [comment]
> @mykaul we discussed it shortly yesterday with @ScyllaPiotr. We believe that making sure that backup/restore works correctly and covering it by proper tests is of high priority from the perspective of reliability and data safety.


Which of the two alternatives presented by @Michal-Leszczynski are you planning to use?
>  - make it safe to use CQL for operating on alternator tables
>  - change SM backup/restore procedure, so that it uses alternator API instead of CQL where CQL is unsafe to use

**Comment by swasik** [comment]
> Which of the two alternatives presented by [@Michal-Leszczynski](https://github.com/Michal-Leszczynski) are you planning to use?

I agree with @nyh arguments that making CQL compatible with Alternator tables. This would be big design decision change against what was decided when Alternator was designed and would make CQL implementation more complex. I think that we should have proper support for Alternator in SM instead.

**Comment by nyh** [comment]
Adding to the list of Alternator-only features that can't currently be replicated by a CQL "CREATE TABLE" and friends - see 
https://github.com/scylladb/scylladb/issues/19112#issuecomment-2149953145 - there is now a fifth one:

5. Starting with pull request https://github.com/scylladb/scylladb/pull/21989 "implement UpdateTable operation to add or delete GSI", Alternator's GSI now use an even more unusual form of materialized views, which uses a "computed column" that knows how to read Alternator-encoded non-key attributes (`extract_from_attrs_column_computation`). Needless to say, CQL "CREATE MATERIALIZED VIEW" can't create such a materialized view currently. It could, if we add a "WITH ALTERNATOR" option to it or some new syntax to create computed columns - but today it just can't. To be honest I also haven't checked what "DESCRIBE SCHEMA" outputs on such a materialized view.

**Comment by ScyllaPiotr** [comment]
> Which of the two alternatives presented by [@Michal-Leszczynski](https://github.com/Michal-Leszczynski) are you planning to use?
> 
> > * make it safe to use CQL for operating on alternator tables
> > * change SM backup/restore procedure, so that it uses alternator API instead of CQL where CQL is unsafe to use

The latter.

**Comment by ScyllaPiotr** [comment]
@Michal-Leszczynski following https://github.com/scylladb/scylla-manager/issues/4509 closed, is this issue ready to be closed too (as not planned)?

**Comment by Michal-Leszczynski** [comment]
I think it can be closed as not planned - we went with other direction of using specific alternator APIs for backing up alternator schema instead of adjusting DESC SCHEMA WITH INTERNALS output to fully cover alternator schema.
